Servers/Administration/OJS.Servers.Administration/Controllers/ProblemResourcesController.cs

```diff
@@ -0,0 +1,186 @@
+        => base.PostEdit(entityDict, new FormFilesContainer(file));
+
+    public async Task<IActionResult> Download([FromQuery] IDictionary<string, string> complexId)
+    {
+        var id = this.GetEntityIdFromQuery(complexId);
+
+        var resource = await this.problemResourcesData.OneByIdTo<ProblemResourceDownloadServiceModel>(id);
+
+        await this.problemResourcesDownloadValidation
+            .GetValidationResult(resource)
+            .VerifyResult();
+
+        var file = resource?.File;
+
+        if (file == null)
+        {
+            return this.NotFound();
+        }
+
+        var contentType = this.contentTypes.GetByFileExtension(resource!.FileExtension);
+        var fileName = $"Resource-{resource.Id}-{resource.ProblemName.Replace(" ", string.Empty)}.{resource.FileExtension}";
+
+        return this.File(file, contentType, fileName);
+    }
+
+    protected override IEnumerable<Func<ProblemResource, ProblemResource, AdminActionContext, ValidatorResult>>
+        EntityValidators
+        => this.problemResourceValidatorsFactory.GetValidators();
+
+    protected override IEnumerable<Func<ProblemResource, ProblemResource, AdminActionContext, Task<ValidatorResult>>>
+        AsyncEntityValidators
+        => this.problemResourceValidatorsFactory.GetAsyncValidators();
+
+    protected override IEnumerable<GridAction> CustomActions
+        => new[]
+        {
+            new GridAction { Action = nameof(Download) }
+        };
+
+    protected override object GetDefaultRouteValuesForPostEntityFormRedirect(
+        ProblemResource newEntity)
+        => new Dictionary<string, string>
+        {
+            { ProblemIdKey, newEntity.ProblemId.ToString() },
+        };
+
+    protected override async Task<IEnumerable<FormControlViewModel>> GenerateFormControlsAsync(
+        ProblemResource entity,
+        EntityAction action,
+        IDictionary<string, string> entityDict,
+        IDictionary<string, Expression<Func<object, bool>>> complexOptionFilters)
+    {
+        var formControls = await base.GenerateFormControlsAsync(entity, action, entityDict, complexOptionFilters)
+            .ToListAsync();
+
+        var problemId = entityDict.TryGetEntityId<Problem>();
+
+        if (problemId == null)
+        {
+            throw new Exception($"A valid ProblemId must be provided to be able to {action} a Problem Resource.");
+        }
+
+        var problemInput = formControls.First(fc => fc.Name == nameof(ProblemResource.Problem));
+        problemInput.Value = problemId;
+        problemInput.IsReadOnly = true;
+
+        var orderByInput = formControls.First(fc => fc.Name == nameof(entity.OrderBy));
+        orderByInput.Value = await this.GetNewOrderBy(problemId.Value);
+
+        formControls.Add(new FormControlViewModel
+        {
+            Name = AdditionalFormFields.File.ToString(),
+            Type = typeof(IFormFile),
+        });
+
+        return formControls;
+    }
```

Comment:
The whole method can be extracted into a method. This will:
1. Give name to this logic
2. Make the actual `GenerateFormControlsAsync` override simpler:

```
protected override async Task<IEnumerable<FormControlViewModel>> GenerateFormControlsAsync(
        ProblemResource entity,
        EntityAction action,
        IDictionary<string, string> entityDict,
        IDictionary<string, Expression<Func<object, bool>>> complexOptionFilters)
    {
        var formControls = await base.GenerateFormControlsAsync(entity, action, entityDict, complexOptionFilters)
            .ToListAsync();
        return formControls.Append(this.DoTheThing());
}
```